File: Top View of Binary Tree.cpp
/* A binary tree node structure
struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};
*/

class Solution {
public:
    // Function to return a list of nodes visible from the top view from left to right
    static vector<int> topView(Node* root) {
        vector<int> arr;
        if (root == nullptr) return arr;

        map<int, int> hm; // map to store horizontal distance -> node value
        queue<Node*> q;
        queue<int> distQ;

        q.push(root);
        distQ.push(0);

        while (!q.empty()) {
            Node* top = q.front();
            q.pop();
            int topDist = distQ.front();
            distQ.pop();

            if (hm.find(topDist) == hm.end()) {
                hm[topDist] = top->data;
            }

            if (top->left) {
                q.push(top->left);
                distQ.push(topDist - 1);
            }

            if (top->right) {
                q.push(top->right);
                distQ.push(topDist + 1);
            }
        }

        for (auto& pair : hm) {
            arr.push_back(pair.second);
        }

        return arr;
    }
};

File: Top View of Binary Tree.java
/*
class Node{
    int data;
    Node left;
    Node right;
    Node(int data){
        this.data = data;
        left=null;
        right=null;
    }
}
*/

class Solution {
    static ArrayList<Integer> topView(Node root) {
        ArrayList<Integer> arr = new ArrayList<>();
        Map<Integer, Integer> hm = new TreeMap<>();
        
        Queue<Node> q = new LinkedList<>();
        Queue<Integer> distQ = new LinkedList<>();
        q.offer(root);
        distQ.offer(0);
        int dist = 0;
        hm.put(dist, root.data);
        
        while(!q.isEmpty()){
            Node top = q.poll();
            int topDist = distQ.poll();
            
            if(!hm.containsKey(topDist)){
                hm.put(topDist, top.data);
            }
            
            if(top.left != null){
                q.offer(top.left);
                distQ.offer(topDist - 1);
            }
            if(top.right != null){
                q.offer(top.right);
                distQ.offer(topDist + 1);
            }
        }
        
        for(Integer i : hm.values()) arr.add(i);
        return arr;
    }
}